### PR TITLE
Consolidate Visual C++ and GCC subprojects and set up CI

### PR DESCRIPTION
**Pull request by vchlin**
- Consolidate Visual C++ and GCC subprojects
- Expand CMake presets to add support for Linux
- Documentation update
- Add a GitHub action file to set up CI
- Speed up CI builds by caching Vcpkg archives (reduce build time from >30 minutes to 1~2 minutes)

### CODE REVIEW

File: gcc-client-reference/optimizer.pb.h
Comment (Groostav) [inline gcc-client-reference/optimizer.pb.h:1]: Regarding generated code deletion: I think this is fine so long as people can easily run cmake. Our previous partners have simply copy-pasted this code into their environment to run it --they may have had to comment out some `#ERROR` stateements because they couldnt get exactly the same version, which is bad for the reasons you mentioned in the previous PR.

I would like to run this before approving it, but i am always glad to get generated code out of the repo.

**Review by Groostav — CHANGES_REQUESTED** [review]
This looks good vincent

Before i approve it I'm going to ask cenk to check it out and build it on his machine.

If he cant I'll come back and issue an approval and assume that the github actions success is sufficient. 

Along those lines, how is github actions when it fails? Does it give you useful logs for debugging? If you introduce a compiler error and push it, does github actions take the compilation errors and forward them to a failure report screen?

Anyways, only a few things i've asked you to do here. Give me some comments and hopefuly before you do ill get back to you regarding approval

File: .github/workflows/cmake-multi-platform.yml
Comment (vchlin) [reply .github/workflows/cmake-multi-platform.yml:7]: It means GitHub actions run when the `master` branch is pushed or when pull requests against `master` are created (or updated). I can enable them for all branches if we think that's necessary.

File: .github/workflows/cmake-multi-platform.yml
Comment (vchlin) [reply .github/workflows/cmake-multi-platform.yml:11]: It's how GitHub actions handle multi-platform testing. This expands into the combinations of platforms / configs you specify.

File: .github/workflows/cmake-multi-platform.yml
Comment (vchlin) [reply .github/workflows/cmake-multi-platform.yml:26]: Yeah one run each. A preset here is a compiler toolchain plus a set of features, flags, env etc. to use when building the project. This is entirely user-defined and  is just one dimension of the matrix. `os` is a built-in provided by the build matrix though.

File: cpp-client-reference/CMakePresets.json
Comment (vchlin) [reply cpp-client-reference/CMakePresets.json:29]: Yeah, this is just MSVC.

File: cpp-client-reference/CMakePresets.json
Comment (vchlin) [reply cpp-client-reference/CMakePresets.json:26]: Yeah. This is just an "abstract" config that other configs inherit from. When it's hidden, the user can't ask to use it directly.

File: cpp-client-reference/CMakePresets.json
Comment (vchlin) [reply cpp-client-reference/CMakePresets.json:52]: Yeah. This is just CMake stuff and doesn't have to with GitHub.

File: cpp-client-reference/CMakePresets.json
Comment (vchlin) [reply cpp-client-reference/CMakePresets.json:102]: This is by convention. I believe this is now the recommended way to supply multiple supported configs with CMake.

File: gcc-client-reference/optimizer.pb.h
Comment (vchlin) [reply gcc-client-reference/optimizer.pb.h:1]: Yeah valid concern. I am thinking we can provide them with a release containing generated files in that case. What do you think?

**Comment by vchlin** [comment]
> This looks good vincent
> 
> Before i approve it I'm going to ask cenk to check it out and build it on his machine.
> 
> If he cant I'll come back and issue an approval and assume that the github actions success is sufficient.
> 
> Along those lines, how is github actions when it fails? Does it give you useful logs for debugging? If you introduce a compiler error and push it, does github actions take the compilation errors and forward them to a failure report screen?
> 
> Anyways, only a few things i've asked you to do here. Give me some comments and hopefuly before you do ill get back to you regarding approval

Thanks for looking, Geoff. So GitHub actions show everything that goes to stdout / stderr, which include all compilation errors. This is fine for our own code. One issue though is that Vcpkg likes to put its build errors in a log file and tells you to go look. But AFAICT there's no easy way to access this file. I'll probably add an action step that `cat`s the log when it fails.

File: .github/workflows/cmake-multi-platform.yml
Comment (vchlin) [reply .github/workflows/cmake-multi-platform.yml]: Good point. Fixed. 

I think in practise this should be fine though. GitHub actions set up the workspace for us and I'd think they wouldn't make things difficult for people.

File: .github/workflows/cmake-multi-platform.yml
Comment (vchlin) [reply .github/workflows/cmake-multi-platform.yml]: Fixed.

**Comment by cferidun** [comment]
I am able to compile (VS2022) and run cpp reference client with OASIS Volition 2023.

**Comment by vchlin** [comment]
> I am able to compile (VS2022) and run cpp reference client with OASIS Volition 2023.

Thanks for looking, Cenk! Can I go ahead and merge this?

**Comment by cferidun** [comment]
Yes, you can. Thank you.

**Pull request merged**
